Detectors/TPC/calibration/src/CalibdEdx.cxx

```diff
@@ -11,91 +11,296 @@
+void CalibdEdx::finalize()
 {
-  auto& cont = getSlots();
-  auto& slot = front ? cont.emplace_front(tstart, tend) : cont.emplace_back(tstart, tend);
+  const float entries = minStackEntries();
+  mCalib.clear();
+
+  TLinearFitter fitter(2);
+
+  // Choose the fit dimension based on the available statistics
+  if (entries >= mFitCuts[2]) {
+    fitter.SetFormula("1 ++ x ++ x*x ++ y ++ x*y ++ y*y");
+    mCalib.setDims(2);
+  } else if (entries >= mFitCuts[1]) {
+    fitter.SetFormula("1 ++ x ++ x*x");
+    mCalib.setDims(1);
+  } else {
+    fitter.SetFormula("1");
+    mCalib.setDims(0);
+  }
+  LOGP(info, "Fitting {}D dE/dx correction for GEM stacks", mCalib.getDims());
 
-  auto container = std::make_unique<dEdxHistos>(mNBins, mCuts);
-  container->setApplyCuts(mApplyCuts);
+  // if entries bellow minimum threshold, integrate all sectors
+  if (mCalib.getDims() == 0 || entries >= mFitCuts[0]) {
+    fitHist(mHist, mCalib, fitter);
+  } else {
+    LOGP(info, "Integrating GEM stacks sectors in dE/dx correction due to low statistics");
 
-  slot.setContainer(std::move(container));
-  return slot;
+    // get mean of each GEM stack
+    CalibdEdxCorrection meanCorr{};
+    meanCorr.setDims(0);
+    TLinearFitter meanFitter(0);
+    meanFitter.SetFormula("1");
+    fitHist(mHist, meanCorr, meanFitter);
+
+    // get highier dimension corrections with projected sectors
+    fitHist(mHist, mCalib, fitter, &meanCorr);
+  }
 }
 
-void CalibdEdx::enableDebugOutput(std::string_view fileName)
+float CalibdEdx::minStackEntries() const
 {
-  mDebugOutputStreamer = std::make_unique<o2::utils::TreeStreamRedirector>(fileName.data(), "recreate");
+  // sum over the dEdx bins to get the number of entries per stack
+  auto projection = bh::algorithm::project(mHist, std::vector<int>{Axis::Sector, Axis::Stack});
+  auto dEdxCounts = bh::indexed(projection);
+  // find the stack with the least number of entries
+  auto min_it = std::min_element(dEdxCounts.begin(), dEdxCounts.end());
+  // the count is doubled since we sum qMax and qTot entries
+  return static_cast<float>(*min_it / 2);
 }
 
-void CalibdEdx::disableDebugOutput()
+bool CalibdEdx::hasEnoughData(float minEntries) const
 {
-  // This will call the TreeStream destructor and write any stored data.
-  mDebugOutputStreamer.reset();
+  return minStackEntries() >= minEntries;
 }
 
-void CalibdEdx::finalizeDebugOutput() const
+TH2F CalibdEdx::getRootHist(const std::vector<int>& projected_axis) const
 {
-  if (mDebugOutputStreamer) {
-    LOG(INFO) << "Closing dump file";
-    mDebugOutputStreamer->Close();
+  const float lower = mHist.axis(0).begin()->lower();
+  const float upper = mHist.axis(0).end()->lower();
+
+  auto projectedHist = getHist(projected_axis);
+
+  const int nBins = mHist.axis(Axis::dEdx).size();
+  const int nHists = projectedHist.size() / nBins;
+
+  TH2F rootHist("", "", nHists, 0, nHists, nBins, lower, upper);
```

Comment:
Why no name and title? Can this be deduced from the axes names? Also the axis titles could be set I think.

Reply:
Sure, I should set a name and title, but the axis will be tricky to set because I'm doing a 2D unrolling of the multi D hist and I also can apply projections to it.
I am thinking about removing it, to be honest. I keep this method so I can plot 2D hists easily, but since now I can dump the hist content to a tree maybe it is not useful anymore.

Reply:
I think you can keep it. But just give it some name, e.g. "hdEdxProjection" or so.